File: 0000-0999/0413.ArithmeticSlices.go
func numberOfArithmeticSlices(A []int) int {
    n := len(A)
    cnt := 0
    // O(n) + O(n)
    dp := make([]int, n)
    for i := 2; i < n; i++ {
        if A[i]-A[i-1] == A[i-1]-A[i-2] {
            dp[i] = dp[i-1] + 1
            cnt += dp[i]
        }
    }
    /* O(n^2) + O(1)
    for i := 1; i < n-1; i++ {
        for j := i+1; j < n; j++ {
            if A[j]-A[j-1] == A[j-1]-A[j-2] {
                cnt++
            } else {
                break
            }
        }
    }*/
    return cnt
}

File: 0000-0999/0413.ArithmeticSlices.py
class Solution(object):
    def numberOfArithmeticSlices(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        cur = ans = 0
        for i in xrange(2, len(A)):
            if A[i]-A[i-1] == A[i-1]-A[i-2]:
                cur += 1
                ans += cur
            else:
                cur = 0
        return ans
        '''
        n = len(A)
        ans = 0
        for l in xrange(n-2):
            d = A[l+1] - A[l]
            for r in xrange(l+2, n):
                if A[r]-A[r-1] == d:
                    ans += 1
                else:
                    break
        return ans
        '''